tests/page_test.go

```diff
@@ -703,6 +706,233 @@ func TestPageClose(t *testing.T) {
+			consoleFn: "() => console.group()",
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "startGroup" &&
+					cm.Text() == "console.group" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.groupCollapsed",
+			consoleFn: "() => console.groupCollapsed()",
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "startGroupCollapsed" &&
+					cm.Text() == "console.groupCollapsed" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.groupEnd",
+			consoleFn: "() => console.groupEnd()",
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "endGroup" &&
+					cm.Text() == "console.groupEnd" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.assert",
+			consoleFn: "() => console.assert(2 == 3)", // Only writes to console if assertion is false
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "assert" &&
+					cm.Text() == "console.assert" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.count (default label)",
+			consoleFn: "() => console.count()", // default label
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "count" && //nolint:goconst
+					cm.Text() == "default: 1" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.count",
+			consoleFn: "() => console.count()", // default label
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "count" &&
+					cm.Text() == "default: 1" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.count",
+			consoleFn: "() => console.count('k6')",
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "count" &&
+					cm.Text() == "k6: 1" &&
+					cm.Page().URL() == blankPage
+			},
+		},
+		{
+			name:      "on console.time",
+			consoleFn: "() => { console.time('k6'); console.timeEnd('k6'); }",
+			assertFn: func(cm *common.ConsoleMessage) bool {
+				return cm.Type() == "timeEnd" && strings.HasPrefix(cm.Text(), "k6: 0.0") &&
+					cm.Page().URL() == blankPage
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Use withSkipClose() opt as we will close it
+			// manually to force the page.TaskQueue closing
+			tb := newTestBrowser(t, withSkipClose())
+			p := tb.NewPage(nil)
+
+			var assertCalled, assertFail bool
+
+			eventHandler := func(mssg api.ConsoleMessage) error { //nolint:unparam
+				assertCalled = true
+				cm, ok := mssg.(*common.ConsoleMessage)
+				if !ok || !tc.assertFn(cm) {
+					assertFail = true
+				}
+				return nil
+			}
+
+			err := tb.vu.Loop.Start(func() error {
+				r := tb.vu.Loop.RegisterCallback()
+				go func() {
+					// Wait some time for the browser to
+					// generate the onConsoleAPICalled event
+					time.Sleep(200 * time.Millisecond)
```

Comment:
This can be racy (although rare, but it can be environment-specific). What do you think about using a channel instead? Let the `eventHandler` close the chan and the goroutine wait for it to be closed before closing the task queue.

Reply:
I had a similar implementation previously, but using a `sync.WaitGroup`. The problem is that in that case the test could hang if the handler is not called for some reason, so decided to change it to use the current implementation, which pretty much waits fro a TO of 200ms. Do you think it would be better to wait indefinitely?

Reply:
I think it would be better to let the tests time out due to `go test` time limit if the handlers do not get called rather than to have possible flaky tests. Have you considered using a timeout context instead?

Reply:
Changed the test to wait for both handlers to be called or up to a more than reasonable timeout. Also added testing for messages multiplexing. Nice catch finding an issue with that!
See https://github.com/grafana/xk6-browser/pull/1006/commits/e4f4a30efb26d911a3704390cd547cfdbc48a237

Reply:
Nice 👍 I suggest heeding this:

[WaitGroup.Add](https://pkg.go.dev/sync#WaitGroup.Add):

> Note that calls with a positive delta that occur when the counter is zero must happen before a Wait. Calls with a negative delta, or calls with a positive delta that start when the counter is greater than zero, may happen at any time. Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for. If a WaitGroup is reused to wait for several independent sets of events, new Add calls must happen after all previous Wait calls have returned. See the WaitGroup example.

Since we're calling `wg.Add` **after** `wg.Wait()`, we might get across with unintended behavior:

```go
err := tb.vu.Loop.Start(func() error {
	// ...
	go func() {
		// ...
		go func() {
			wg.Wait() // <----
			close(done)
		}()

		select {
		case <-done:
		case <-time.After(testTO):
			assertTO = true
		}

		r(func() error {
			tb.Close()
			return nil
		})
	}()
	wg.Add(1) // <--- AFTER
	err := p.On("console", eventHandlerOne)
	if err != nil {
		return err
	}
	wg.Add(1) // <--- AFTER
	err = p.On("console", eventHandlerTwo)
	// ...
})
```

Reply:
Yep, just realized this, fixed in [f72e9c1](https://github.com/grafana/xk6-browser/pull/1006/commits/f72e9c1269cb84531a0c8f9e9126cfbc6f8a843f).

Reply:
It's not unexpected to deal with this stuff when we incorporate [event-loop handling in the business logic](https://github.com/grafana/xk6-browser/pull/1006#discussion_r1305513979) :(